app/src/main/java/org/schabi/newpipe/util/InfoCache.java

```diff
@@ -52,25 +55,27 @@ public static InfoCache getInstance() {
         return instance;
     }
 
+    @Nullable
     public Info getFromKey(int serviceId, @NonNull String url) {
         if (DEBUG) Log.d(TAG, "getFromKey() called with: serviceId = [" + serviceId + "], url = [" + url + "]");
         synchronized (lruCache) {
             return getInfo(lruCache, keyOf(serviceId, url));
         }
     }
 
-    public void putInfo(@NonNull Info info) {
+    public void putInfo(int serviceId, @NonNull String url, @NonNull Info info) {
         if (DEBUG) Log.d(TAG, "putInfo() called with: info = [" + info + "]");
-        synchronized (lruCache) {
-            final CacheData data = new CacheData(info, DEFAULT_TIMEOUT_HOURS, TimeUnit.HOURS);
-            lruCache.put(keyOf(info), data);
+
+        final long expirationMillis;
+        if (info.getServiceId() == SoundCloud.getServiceId()) {
```

Comment:
@theScrabi
I was hoping to receive some more review and discussions before having the PR merged, since there might be parts that would lead to tech debts.
For example, this part where different services can have different cache expiration. SoundCloud was added here since their API seems to fail sporadically and would cause the `StreamInfo` to have bad stream urls and need to be reloaded. 

This probably should be part of the extractor, or at least `ServiceHelper`. Please consider removing this.

Reply:
I see, I merged early, so now we have to take care about this in the aftermath.

> For example, this part where different services can have different cache expiration. SoundCloud was added here since their API seems to fail sporadically and would cause the `StreamInfo` to have bad stream urls and need to be reloaded.

So you`d suggest a new field in `StreamInfo` that tells the URls expiration time? If so we should add this.

>  Please consider removing this.

You mean the part that checks for SoundCloud? Right thats a hack and needs a proper solution.

Reply:
I'm currently working on a PR that lets user wipe the `InfoCache`, so we can remove this check. I've also reduce the default cache time to 1 hr instead of 4. I've also moved that check to `ServiceHelper` for now and reduce the cache time for SoundCloud to 5 minutes.

However, this is not a long term solution, as SoundCloud stream url seems to consistently expire after a few minutes if they are not played, at least from my side. We will need to figure out exactly why this is happening, since it's probably a bad idea just slapping on an arbitrary expiration time for each service.

Reply:
Well we might have similare problems with YouTube as some issues point out. But you are right, we should investigate this.